core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -585,6 +591,23 @@ class ReplicaManager(val config: KafkaConfig,
                     result.info.logStartOffset, result.info.recordErrors.asJava, result.info.errorMessage)) // response status
       }
 
+      actionQueue.add {
+        () =>
+          localProduceResults.foreach {
+            case (topicPartition, result) =>
+              val requestKey = TopicPartitionOperationKey(topicPartition)
+              if (result.info.leaderHWIncremented) {
```

Comment:
It seems that we have to distinguish 3 states here: (1) records not appended due to an error; (2) records appended successfully and HWM advanced; (3) records appended successfully and HWM not advanced. In case (1), no purgatory needs to be checked.

Reply:
you are right. I miss the exception in ```appendToLocalLog```. Let me revert this change